File: LeetCode/src/designclass/MyCircularQueue.java
package designclass;

public class MyCircularQueue {
    private int[] queue;
    private int capacity;

    private int first;

    private  int size;
    public MyCircularQueue(int k) {
        queue = new int[k];
        capacity = k;
        first = 0;
        size = 0;
    }

    public boolean enQueue(int value) {

        if (isFull()) {
            return false;
        } else {
            queue[(first + size) % capacity] = value;
            size++;
            return true;
        }
    }

    public boolean deQueue() {
        if (isEmpty()) {
            return false;
        } else {
            first = (first + 1) % capacity;
            size--;
            return true;
        }
    }

    public int Front() {
        return isEmpty() ? -1 : queue[first];
    }

    public int Rear() {
        return isEmpty() ? -1 : queue[(first + size - 1) % capacity];
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return size == capacity;
    }

    public static void main(String[] args) {
        MyCircularQueue circularQueue = new MyCircularQueue(3); // set the size to be 3
        System.out.println(circularQueue.enQueue(1)); // return true
        System.out.println(circularQueue.enQueue(2)); // return true
        System.out.println(circularQueue.enQueue(3)); // return true
        System.out.println(circularQueue.enQueue(4)); // return false, the queue is full
        System.out.println(circularQueue.Rear()); // return 3
        System.out.println(circularQueue.isFull()); // return true
        System.out.println(circularQueue.deQueue()); // return true
        System.out.println(circularQueue.enQueue(4)); // return true
        System.out.println(circularQueue.Rear()); // return 4
    }
}

File: LeetCode_Python/designclass/MyCircularQueue.py
class MyCircularQueue:
    def __init__(self, k: int):
        self.queue = [0] * k
        self.head = 0
        self.tail = 0
        self.size = 0
        self.capacity = k

    def enQueue(self, value: int) -> bool:
        if self.size == self.capacity:
            return False
        self.queue[self.tail] = value
        self.tail = (self.tail + 1) % self.capacity
        self.size += 1
        return True

    def deQueue(self) -> bool:
        if self.size == 0:
            return False
        self.head = (self.head + 1) % self.capacity
        self.size -= 1
        return True

    def Front(self) -> int:
        if self.size == 0:
            return -1
        return self.queue[self.head]

    def Rear(self) -> int:
        if self.size == 0:
            return -1
        return self.queue[(self.tail - 1 + self.capacity) % self.capacity]

    def isEmpty(self) -> bool:
        return self.size == 0

    def isFull(self) -> bool:
        return self.size == self.capacity


# Time Complexity:
# - O(1) for all operations (enQueue, deQueue, Front, Rear, isEmpty, isFull)
# Space Complexity:
# - O(k) where k is the size of the circular queue

if __name__ == "__main__":
    solution = MyCircularQueue(3)

    print("Expected Output: True")
    print("Actual Output:", solution.enQueue(1))

    print("Expected Output: True")
    print("Actual Output:", solution.enQueue(2))

    print("Expected Output: True")
    print("Actual Output:", solution.enQueue(3))

    print("Expected Output: False")
    print("Actual Output:", solution.enQueue(4))

    print("Expected Output: 1")
    print("Actual Output:", solution.Front())

    print("Expected Output: 3")
    print("Actual Output:", solution.Rear())

    print("Expected Output: True")
    print("Actual Output:", solution.deQueue())

    print("Expected Output: True")
    print("Actual Output:", solution.enQueue(4))
